From da5b29291dd7d2f6242cae5a83139246c1d4fa5b Mon Sep 17 00:00:00 2001
From: Dmitry R <rdmitry0911@gmail.com>
Date: Thu, 1 Oct 2026 00:18:27 +0300
Subject: [PATCH manager 1/3] pveproxy: serve the page of a WebRTC console
 viewer

Like noVNC (novnc-pve) and xterm.js (pve-xtermjs), a WebRTC console
viewer is a page from a separate package: '?console=kvm&webrtc=1'
renders /usr/share/pve-webrtc-console/index.html.tpl, with its assets
under /webrtc/.  The viewer uses the API call webrtcproxy of the VM,
so it needs no port or authentication of its own.

If no viewer is installed, the request is answered with 501 and a
short explanation.

Signed-off-by: Dmitry R <rdmitry0911@gmail.com>
---
 PVE/Service/pveproxy.pm | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/PVE/Service/pveproxy.pm b/PVE/Service/pveproxy.pm
index dfdd014c..8ecb106a 100755
--- a/PVE/Service/pveproxy.pm
+++ b/PVE/Service/pveproxy.pm
@@ -54,6 +54,7 @@ my $basedirs = {
     i18n => '/usr/share/pve-i18n',
     manager => '/usr/share/pve-manager',
     novnc => '/usr/share/novnc-pve',
+    webrtc => '/usr/share/pve-webrtc-console',
     yew_mobile => '/usr/share/pve-yew-mobile-gui',
     i18n_yew => '/usr/share/pve-yew-mobile-i18n',
     widgettoolkit => '/usr/share/javascript/proxmox-widget-toolkit',
@@ -91,6 +92,7 @@ sub init {
     add_dirs($dirs, '/pwt/css/' => "$basedirs->{widgettoolkit}/css/");
     add_dirs($dirs, '/pwt/images/' => "$basedirs->{widgettoolkit}/images/");
     add_dirs($dirs, '/pwt/themes/' => "$basedirs->{widgettoolkit}/themes/");
+    add_dirs($dirs, '/webrtc/' => "$basedirs->{webrtc}/");
     add_dirs($dirs, '/xtermjs/' => "$basedirs->{xtermjs}/");
     add_dirs($dirs, '/yew-mobile/', "$basedirs->{yew_mobile}/");
     add_dirs($dirs, '/yew-mobile/i18n/', "$basedirs->{i18n_yew}/");
@@ -249,6 +251,7 @@ sub get_index {
 
     my $novnc = defined($args->{console}) && $args->{novnc};
     my $xtermjs = defined($args->{console}) && $args->{xtermjs};
+    my $webrtc = defined($args->{console}) && $args->{webrtc};
 
     my $langfile = -f "$basedirs->{i18n}/pve-lang-$lang.js" ? 1 : 0;
 
@@ -295,6 +298,18 @@ sub get_index {
         $dir = $basedirs->{novnc};
     } elsif ($xtermjs) {
         $dir = $basedirs->{xtermjs};
+    } elsif ($webrtc) {
+        # the viewer comes with the node's WebRTC console service, a separate package
+        if (!-f "$basedirs->{webrtc}/index.html.tpl") {
+            my $headers = HTTP::Headers->new(Content_Type => "text/plain; charset=utf-8");
+            return HTTP::Response->new(
+                501,
+                "Not Implemented",
+                $headers,
+                "The WebRTC console viewer is not installed on this node.\n",
+            );
+        }
+        $dir = $basedirs->{webrtc};
     } elsif ($mobile) {
         $dir = "$basedirs->{yew_mobile}";
     }
-- 
2.43.0

